### PR TITLE
feat: add source-dev BEAM operating model

### PR DESCRIPTION
**Pull request by najibninaba**
## Intent

Implement the OpenSpec source-dev BEAM operating model for Orchard so source development can run split controller and node-agent BEAM roles with explicit environment parsing and safe defaults. The change wires source-dev BEAM config, adds bin/dev-controller and bin/dev-node-agent bootstrap behavior, preserves all-in-one bin/dev on the gRPC runtime default, rejects BEAM-specific distributed runtime flags in all-in-one mode, adds cookie generation and validation with race-safe default cookie creation, validates EPMD and distribution ports, and keeps diagnostics secret-safe. It also adds regressions proving runtime dispatch, scheduler, Console Runtime, Console Nodes, and node-agent endpoint mapping do not silently fall back when runtime targets are unreachable. Documentation and OpenSpec tasks were updated, with real two-Mac source-dev smoke evidence intentionally deferred by leaving tasks 5.4 and 5.5 unchecked.

## What Changed

- Adds explicit source-dev BEAM Runtime Endpoint mode for split controller and node-agent roles, wiring BEAM target parsing, BeamClient config, guardrails, and all-in-one `bin/dev` rejection while preserving gRPC as the default path.
- Adds shared BEAM bootstrap support for `bin/dev-controller` and `bin/dev-node-agent`, including named-node launch args, cookie creation and validation, EPMD preflight, IPv4/interface binding, bounded distribution ports, and secret-safe diagnostics.
- Hardens scheduler, dispatch, Console, runtime endpoint mapping, docs, OpenSpec tasks, and regression coverage so BEAM target failures do not silently fall back to gRPC.

## Risk Assessment

⚠️ Medium: No material merge-blocking issues were found, but this is a nontrivial source-dev startup and networking path with real distributed smoke evidence still deferred.

## Testing

The bootstrap contract passed; the first Mix run was blocked by mise trust, then the same focused test set passed with an environment-only trust override. I also produced CLI dry-run evidence and a no-fallback trace summary; no screenshot was captured because the changed Console files are tests only, not UI source or layout changes.

<details>
<summary>Evidence: Source-dev BEAM CLI transcript</summary>

```text
Source-dev BEAM CLI evidence

Actual all-in-one bin/dev with ORCHARD_RUNTIME_ENDPOINT_TRANSPORT=beam
error: all-in-one bin/dev does not support BEAM source-dev mode yet
error: use bin/dev-controller and bin/dev-node-agent for ORCHARD_RUNTIME_ENDPOINT_TRANSPORT=beam
mix_called_before_rejection=no

Split-role controller dry-run
==> Runtime endpoint transport: beam
==> BEAM node name: orchard_controller@127.0.0.1
==> BEAM cookie file: /var/folders/xy/bmzx3f853zd1tdvd11hgc0gm0000gn/T/no-mistakes-evidence/01KW24H36B7F8K1T9SSFQNY102/source-dev-beam-cli-sandbox/entrypoint-repo/tmp/dev/beam.cookie
==> BEAM EPMD port: 4369
==> BEAM distribution port range: 52171..52171
==> Ensuring dev database...
==> Starting Orchard dev controller (role: controller)
==> HTTP endpoint: 127.0.0.1:4000
==> Runtime client targets: 127.0.0.1:50071

controller_mix_calls
mix called: ecto.create --quiet
mix called: ecto.migrate --quiet
controller_iex_invocation
ERL_EPMD_ADDRESS=127.0.0.1
--name orchard_controller@127.0.0.1 --erl -kernel inet_dist_use_interface {127,0,0,1} inet_dist_listen_min 52171 inet_dist_listen_max 52171 -S mix phx.server

Split-role node-agent dry-run
warning: ORCHARD_WORKER_EXECUTABLE is not executable: /var/folders/xy/bmzx3f853zd1tdvd11hgc0gm0000gn/T/no-mistakes-evidence/01KW24H36B7F8K1T9SSFQNY102/source-dev-beam-cli-sandbox/entrypoint-repo/native/orchard_worker_mlx/bin/orchard-worker-mlx
==> Runtime endpoint transport: beam
==> BEAM node name: orchard_node_agent@127.0.0.1
==> BEAM cookie file: /var/folders/xy/bmzx3f853zd1tdvd11hgc0gm0000gn/T/no-mistakes-evidence/01KW24H36B7F8K1T9SSFQNY102/source-dev-beam-cli-sandbox/entrypoint-repo/tmp/dev/beam.cookie
==> BEAM EPMD port: 4369
==> BEAM distribution port range: 52172..52172
==> Starting Orchard dev node-agent (role: node-agent)
==> gRPC endpoint: 127.0.0.1:50071
==> Worker executable: /var/folders/xy/bmzx3f853zd1tdvd11hgc0gm0000gn/T/no-mistakes-evidence/01KW24H36B7F8K1T9SSFQNY102/source-dev-beam-cli-sandbox/entrypoint-repo/native/orchard_worker_mlx/bin/orchard-worker-mlx

node_agent_mix_calls
node_agent_iex_invocation
ERL_EPMD_ADDRESS=127.0.0.1
--name orchard_node_agent@127.0.0.1 --erl -kernel inet_dist_use_interface {127,0,0,1} inet_dist_listen_min 52172 inet_dist_listen_max 52172 -S mix run --no-halt

Cookie and EPMD checks
default_cookie_mode=600
controller_cookie_mode=600
node_agent_cookie_mode=600
controller_cookie_matches_default=yes
node_agent_cookie_matches_default=yes
epmd_calls
epmd -port 4369 -names
epmd -daemon -address 127.0.0.1 -port 4369
epmd -port 4369 -names
epmd -daemon -address 127.0.0.1 -port 4369
cookie_secret_in_transcript=no
```
</details>
<details>
<summary>Evidence: No-fallback regression trace summary</summary>

```text
  * test live runtime renders source-dev BEAM Runtime Endpoint targets with readable labels and stable DOM ids (197.4ms) [L#811]
  * test BEAM Runtime Endpoint dispatch returns BEAM RPC failures without retrying the legacy gRPC target (37.9ms) [L#508]
  * test BEAM Runtime Endpoint dispatch uses the explicit BEAM target and never falls back to the legacy gRPC target (0.3ms) [L#486]
  * test BEAM Runtime Endpoint dispatch returns BEAM stream failures without retrying the legacy gRPC target (9.0ms) [L#533]
  * test cluster_snapshot/0,1 source-dev BEAM target maps use active Runtime Endpoint client and ignore legacy targets (1.5ms) [L#1029]
  * test single-node delegation multi-target BEAM all-probe failure preserves Runtime Endpoint selection (7.8ms) [L#637]
  * test single-node delegation source-dev BEAM target maps emit runtime endpoint schedules without legacy targets (3.7ms) [L#610]
Result: 7 passed, 235 excluded
  * test maps source-dev BEAM target address shape into observation identity (2.0ms) [L#50]
Result: 1 passed, 2 excluded
```
</details>
<details>
<summary>Evidence: No-fallback regression full trace</summary>

```text
==> orchard_controller
Running ExUnit with seed: 673598, max_cases: 1
Excluding tags: [:test]
Including tags: [location: {"test/orchard/dispatch/probe_compatibility_test.exs", 486}, location: {"test/orchard/dispatch/probe_compatibility_test.exs", 508}, location: {"test/orchard/dispatch/probe_compatibility_test.exs", 533}, location: {"test/orchard/scheduler/multi_node_test.exs", 610}, location: {"test/orchard/scheduler/multi_node_test.exs", 637}, location: {"test/orchard/console/runtime_test.exs", 1029}, location: {"test/orchard/console/nodes_live_test.exs", 811}]


OrchardConsole.NodesLiveTest [test/orchard/console/nodes_live_test.exs]
  * test cluster_snapshot exit does not crash LiveView inventory renders correctly despite cluster exit [L#1379]
  * test cluster_snapshot exit does not crash LiveView inventory renders correctly despite cluster exit (excluded) [L#1379]
  * test memory telemetry guardrails memory telemetry alone does not change health labels or summary counts [L#1130]
  * test memory telemetry guardrails memory telemetry alone does not change health labels or summary counts (excluded) [L#1130]
  * test freshness shows freshness row with timestamp after load [L#1295]
  * test freshness shows freshness row with timestamp after load (excluded) [L#1295]
  * test mixed legacy and modern cluster cluster summary counts prompt-token-capable reachable workers [L#1226]
  * test mixed legacy and modern cluster cluster summary counts prompt-token-capable reachable workers (excluded) [L#1226]
  * test live runtime renders map-backed gRPC Runtime Endpoint targets without leaking metadata [L#828]
  * test live runtime renders map-backed gRPC Runtime Endpoint targets without leaking metadata (excluded) [L#828]
  * test safe tokenization telemetry counters HTTP dead mount renders non-zero safe tokenization counters [L#952]
  * test safe tokenization telemetry counters HTTP dead mount renders non-zero safe tokenization counters (excluded) [L#952]
  * test summary strip counts reflect inserted nodes by health [L#716]
  * test summary strip counts reflect inserted nodes by health (excluded) [L#716]
  * test multi-target cluster prompt-token summary denominator excludes unavailable targets [L#1069]
  * test multi-target cluster prompt-token summary denominator excludes unavailable targets (excluded) [L#1069]
  * test safe tokenization telemetry counters raising counter provider fails open and keeps Live Cluster rendering [L#1005]
  * test safe tokenization telemetry counters raising counter provider fails open and keeps Live Cluster rendering (excluded) [L#1005]
  * test runtime unavailable shows per-target error while inventory still renders [L#1248]
  * test runtime unavailable shows per-target error while inventory still renders (excluded) [L#1248]
  * test compatibility warning shown for legacy snapshot (per-target) [L#1084]
  * test compatibility warning shown for legacy snapshot (per-target) (excluded) [L#1084]
  * test memory telemetry guardrails malformed memory telemetry rows fail open as advisory rows [L#1157]
  * test memory telemetry guardrails malformed memory telemetry rows fail open as advisory rows (excluded) [L#1157]
  * test live runtime does not render prompt-token capability badge for unavailable target [L#864]
  * test live runtime does not render prompt-token capability badge for unavailable target (excluded) [L#864]
  * test compatibility warning not shown when runtime is unavailable [L#1108]
  * test compatibility warning not shown when runtime is unavailable (excluded) [L#1108]
  * test memory telemetry guardrails oversized memory telemetry stays bounded at the LiveView seam [L#1171]
  * test memory telemetry guardrails oversized memory telemetry stays bounded at the LiveView seam (excluded) [L#1171]
  * test GET /console/nodes hard mode keeps nodes diagnostics reachable [L#631]
  * test GET /console/nodes hard mode keeps nodes diagnostics reachable (excluded) [L#631]
  * test safe tokenization telemetry counters partial prompt-token counter preserves valid count and defaults token count [L#976]
  * test safe tokenization telemetry counters partial prompt-token counter preserves valid count and defaults token count (excluded) [L#976]
  * test safe tokenization telemetry counters renders non-zero unsafe signals with warning and error tones [L#927]
  * test safe tokenization telemetry counters renders non-zero unsafe signals with warning and error tones (excluded) [L#927]
  * test GET /console/nodes has correct page title [L#671]
  * test GET /console/nodes has correct page title (excluded) [L#671]
  * test memory telemetry guardrails runtime unavailable without memory budgets still renders safely [L#1147]
  * test memory telemetry guardrails runtime unavailable without memory budgets still renders safely (excluded) [L#1147]
  * test same-cycle discovery runtime-first load order surfaces newly discovered node on first render [L#1320]
  * test same-cycle discovery runtime-first load order surfaces newly discovered node on first render (excluded) [L#1320]
  * test mixed legacy and modern cluster cluster summary counts based on reachability not metadata [L#1215]
  * test mixed legacy and modern cluster cluster summary counts based on reachability not metadata (excluded) [L#1215]
  * test safe tokenization telemetry counters counters do not affect health, compatibility, or prompt-token summaries [L#1018]
  * test safe tokenization telemetry counters counters do not affect health, compatibility, or prompt-token summaries (excluded) [L#1018]
  * test sidebar Nodes nav item exists and is active [L#683]
  * test sidebar Nodes nav item exists and is active (excluded) [L#683]
  * test freshness disconnected render shows waiting text without LocalTime hook [L#1304]
  * test freshness disconnected render shows waiting text without LocalTime hook (excluded) [L#1304]
  * test inventory table renders persisted rows [L#739]
  * test inventory table renders persisted rows (excluded) [L#739]
  * test live runtime per-target card renders with stable DOM id [L#804]
  * test live runtime per-target card renders with stable DOM id (excluded) [L#804]
  * test multi-target cluster cluster summary reflects mixed results [L#1059]
  * test multi-target cluster cluster summary reflects mixed results (excluded) [L#1059]
  * test safe tokenization telemetry counters renders process-local observe-only zero counters [L#909]
  * test safe tokenization telemetry counters renders process-local observe-only zero counters (excluded) [L#909]
  * test live runtime shows worker state, backend, and loaded models on full snapshot [L#779]
  * test live runtime shows worker state, backend, and loaded models on full snapshot (excluded) [L#779]
  * test refresh polling via :refresh_nodes updates the page [L#1279]
  * test refresh polling via :refresh_nodes updates the page (excluded) [L#1279]
  * test compatibility warning shown for partial snapshot (per-target) [L#1093]
  * test compatibility warning shown for partial snapshot (per-target) (excluded) [L#1093]
  * test cluster_snapshot exit does not crash LiveView refresh after exit does not crash LiveView [L#1365]
  * test cluster_snapshot exit does not crash LiveView refresh after exit does not crash LiveView (excluded) [L#1365]
  * test live runtime renders observe-only memory telemetry empty state when budgets are absent [L#875]
  * test live runtime renders observe-only memory telemetry empty state when budgets are absent (excluded) [L#875]
  * test compatibility warning memory telemetry does not change compatibility classification [L#1116]
  * test compatibility warning memory telemetry does not change compatibility classification (excluded) [L#1116]
  * test summary strip shows zero-filled counts when no nodes exist [L#706]
  * test summary strip shows zero-filled counts when no nodes exist (excluded) [L#706]
  * test cluster_snapshot raise does not crash LiveView mount succeeds and renders cluster error state [L#1392]
  * test cluster_snapshot raise does not crash LiveView mount succeeds and renders cluster error state (excluded) [L#1392]
  * test multi-target cluster renders one card per target, mixed succ

... [37771 bytes truncated] ...

ng [L#3379]
  * test multi-node scheduling keeps capable-worker preference below live prefix-cache fingerprint matching (excluded) [L#3379]
  * test multi-node scheduling SPEC.md §5.5 reports cold-tier queue capacity from eligible node concurrency [L#1435]
  * test multi-node scheduling SPEC.md §5.5 reports cold-tier queue capacity from eligible node concurrency (excluded) [L#1435]
  * test single-node delegation uses runtime endpoint observation aggregate capacity for active cold target [L#715]
  * test single-node delegation uses runtime endpoint observation aggregate capacity for active cold target (excluded) [L#715]
  * test multi-node scheduling keeps memory admission below historical cache-affinity [L#3030]
  * test multi-node scheduling keeps memory admission below historical cache-affinity (excluded) [L#3030]
  * test single-node delegation delegates to SingleNode when only one target configured [L#528]
  * test single-node delegation delegates to SingleNode when only one target configured (excluded) [L#528]
  * test multi-node scheduling does not let memory admission outrank health [L#3158]
  * test multi-node scheduling does not let memory admission outrank health (excluded) [L#3158]
  * test multi-node scheduling returns dispatch-compatible schedule map [L#3677]
  * test multi-node scheduling returns dispatch-compatible schedule map (excluded) [L#3677]
  * test multi-node scheduling SPEC.md §5.5 excludes active loaded node with invalid matching placement max concurrency [L#1635]
  * test multi-node scheduling SPEC.md §5.5 excludes active loaded node with invalid matching placement max concurrency (excluded) [L#1635]
  * test multi-node scheduling uses cache-affinity as a tie-breaker for otherwise equivalent candidates [L#1988]
  * test multi-node scheduling uses cache-affinity as a tie-breaker for otherwise equivalent candidates (excluded) [L#1988]
  * test multi-node scheduling still schedules deterministically when all workers are legacy [L#3492]
  * test multi-node scheduling still schedules deterministically when all workers are legacy (excluded) [L#3492]
  * test multi-node scheduling keeps prompt-token capability rank-neutral when safe mode is off [L#3306]
  * test multi-node scheduling keeps prompt-token capability rank-neutral when safe mode is off (excluded) [L#3306]
  * test multi-node scheduling prefers prompt-token-capable worker when opt-in flag and safe mode are enabled [L#3252]
  * test multi-node scheduling prefers prompt-token-capable worker when opt-in flag and safe mode are enabled (excluded) [L#3252]
  * test multi-node scheduling prefix cache scoring probes only the selected candidate when enabled [L#2199]
  * test multi-node scheduling prefix cache scoring probes only the selected candidate when enabled (excluded) [L#2199]
  * test multi-node scheduling breaks ties by node_id [L#1861]
  * test multi-node scheduling breaks ties by node_id (excluded) [L#1861]
  * test edge case: all nodes unreachable falls back to SingleNode when all persisted nodes are unreachable [L#3844]
  * test edge case: all nodes unreachable falls back to SingleNode when all persisted nodes are unreachable (excluded) [L#3844]
  * test multi-node scheduling hosted tool capability and readiness data do not change inference ranking [L#1154]
  * test multi-node scheduling hosted tool capability and readiness data do not change inference ranking (excluded) [L#1154]
  * test single-node delegation returns cluster_busy for one live cold target at aggregate capacity [L#765]
  * test single-node delegation returns cluster_busy for one live cold target at aggregate capacity (excluded) [L#765]
  * test multi-node scheduling SPEC.md §7.5 rejects configured BEAM observations before persisting mismatched metadata identity [L#951]
  * test multi-node scheduling SPEC.md §7.5 rejects configured BEAM observations before persisting mismatched metadata identity (excluded) [L#951]
  * test multi-node scheduling tie-only scoring preserves incumbent when both candidates are authoritative resident [L#2446]
  * test multi-node scheduling tie-only scoring preserves incumbent when both candidates are authoritative resident (excluded) [L#2446]
  * test single-node delegation forwards status client options to no-candidate fallback [L#536]
  * test single-node delegation forwards status client options to no-candidate fallback (excluded) [L#536]
  * test multi-node scheduling uses memory admission only as an enabled positive tie-breaker [L#2970]
  * test multi-node scheduling uses memory admission only as an enabled positive tie-breaker (excluded) [L#2970]
  * test edge case: mixed legacy and modern agents schedules to modern node when legacy returns no metadata [L#3885]
  * test edge case: mixed legacy and modern agents schedules to modern node when legacy returns no metadata (excluded) [L#3885]
  * test multi-node scheduling ignores stale cache-affinity placements [L#3574]
  * test multi-node scheduling ignores stale cache-affinity placements (excluded) [L#3574]
  * test multi-node scheduling SPEC.md §5.5 constrains queue admission capacity by node max concurrency [L#1343]
  * test multi-node scheduling SPEC.md §5.5 constrains queue admission capacity by node max concurrency (excluded) [L#1343]
  * test multi-node scheduling keeps memory admission below live prefix-cache fingerprint matching [L#3073]
  * test multi-node scheduling keeps memory admission below live prefix-cache fingerprint matching (excluded) [L#3073]
  * test multi-node scheduling tie-only scoring caps leading tie scoring to two candidates [L#2530]
  * test multi-node scheduling tie-only scoring caps leading tie scoring to two candidates (excluded) [L#2530]
  * test multi-node scheduling tie-only scoring treats recent_fingerprint_only incumbent as comparable non-resident [L#2402]
  * test multi-node scheduling tie-only scoring treats recent_fingerprint_only incumbent as comparable non-resident (excluded) [L#2402]
  * test multi-node scheduling keeps capable-worker preference below historical cache-affinity [L#3332]
  * test multi-node scheduling keeps capable-worker preference below historical cache-affinity (excluded) [L#3332]
  * test multi-node scheduling consumes BEAM runtime endpoint observations without legacy dispatch target [L#901]
  * test multi-node scheduling consumes BEAM runtime endpoint observations without legacy dispatch target (excluded) [L#901]
  * test multi-node scheduling prefix cache scoring normalizes unsupported client versions without affecting selection [L#2822]
  * test multi-node scheduling prefix cache scoring normalizes unsupported client versions without affecting selection (excluded) [L#2822]
  * test single-node delegation multi-target BEAM all-probe failure preserves Runtime Endpoint selection [L#637]
  * test single-node delegation multi-target BEAM all-probe failure preserves Runtime Endpoint selection (7.8ms) [L#637]
  * test single-node delegation source-dev BEAM target maps emit runtime endpoint schedules without legacy targets [L#610]
  * test single-node delegation source-dev BEAM target maps emit runtime endpoint schedules without legacy targets (3.7ms) [L#610]

Finished in 1.9 seconds (0.00s async, 1.9s sync)

Result: 7 passed, 235 excluded
==> orchard_node_agent
Running ExUnit with seed: 673598, max_cases: 1
Excluding tags: [:test]
Including tags: [location: {"test/orchard/node/runtime_endpoint_mapper_test.exs", 50}]


Orchard.Node.RuntimeEndpointMapperTest [test/orchard/node/runtime_endpoint_mapper_test.exs]
  * test maps node-agent status to Runtime Endpoint observation with placement capacity [L#16]
  * test maps node-agent status to Runtime Endpoint observation with placement capacity (excluded) [L#16]
  * test maps prefix-cache score responses to Runtime Endpoint operation results [L#78]
  * test maps prefix-cache score responses to Runtime Endpoint operation results (excluded) [L#78]
  * test maps source-dev BEAM target address shape into observation identity [L#50]
  * test maps source-dev BEAM target address shape into observation identity (2.0ms) [L#50]

Finished in 0.04 seconds (0.04s async, 0.00s sync)

Result: 1 passed, 2 excluded
```
</details>

## Pipeline

Updates from [git push no-mistakes](https://github.com/kunchenguid/no-mistakes)

<details>
<summary>✅ **intent** - passed</summary>

✅ No issues found.

</details>

<details>
<summary>✅ **Rebase** - passed</summary>

✅ No issues found.

</details>

<details>
<summary>🔧 **Review** - 1 issue found → auto-fixed (5) ✅</summary>

- 🚨 `config/dev.exs:160` - BEAM mode can now install multiple explicit Runtime Endpoint targets, but the all-probes-failed multi-target path in MultiNode still falls through to the generic SingleNode gRPC fallback. With two BEAM node agents configured, if both BEAM probes fail the scheduler returns a legacy single-node schedule instead of failing visibly or preserving a BEAM target, which violates the no automatic gRPC fallback contract. Add an explicit non-gRPC Runtime Endpoint fallback/error path for this case and cover the multi-target all-failed scenario.

🔧 Fix: Prevent BEAM multi-target gRPC fallback
3 issues (1 error, 2 warnings) still open:
- 🚨 `config/dev.exs:140` - Explicit BEAM mode is only handled for the controller role. If dev config is evaluated with `ORCHARD_RUNTIME_ENDPOINT_TRANSPORT=beam` and `ORCHARD_SOURCE_DEV_ROLE=all_in_one`, the BEAM target list stays empty, `runtime_endpoint_inference_config` stays empty, and the controller silently uses the legacy gRPC path. Reject `:all_in_one` with BEAM in `config/dev.exs` as well, not only in `bin/dev`.
- ⚠️ `bin/lib/source-dev-beam.sh:130` - The node-agent bootstrap accepts any service starting with `orchard_node_agent`, but the controller-side BEAM target parser and guardrails admit only the exact `orchard_node_agent` service. A node-agent can start as `orchard_node_agent_dev@&lt;ip&gt;` and then be impossible to target through the documented env surface. Decide whether suffixed services are supported, then align bootstrap validation, target parsing, docs, and guardrails.
- ⚠️ `scripts/test-source-dev-beam-bootstrap.sh:1` - This new bootstrap regression suite is standalone and is not invoked by `make test`, `make check-elixir`, or an ExUnit wrapper, so the standard validation path can pass while the shell bootstrap contract regresses. Wire it into a repo validation target or a Mix test that executes the script.

🔧 Fix: Harden source-dev BEAM validation
1 warning still open:
- ⚠️ `bin/lib/source-dev-beam.sh:81` - The BEAM bootstrap validates `ORCHARD_BEAM_NODE_NAME` as `service@ip` and configures guardrails with that host, but the actual IEx launch only constrains the distribution port range. Erlang distribution and EPMD can still bind on wildcard interfaces, so the default `...@127.0.0.1` source-dev run may expose EPMD/distribution outside loopback. Bind distribution with `inet_dist_use_interface` and constrain EPMD to the validated node-name host before launching IEx.

🔧 Fix: Bind source-dev BEAM interfaces
2 warnings still open:
- ⚠️ `bin/lib/source-dev-beam.sh:46` - `ERL_EPMD_ADDRESS` only constrains EPMD when a new daemon is started; if port 4369 already has an EPMD daemon from a prior shell, BEAM mode silently reuses it and can remain wildcard-bound despite the loopback/default-host guard. Add a preflight that starts or verifies an address-constrained EPMD for the selected port, or fails with a clear remediation when an existing daemon cannot be trusted.
- ⚠️ `bin/lib/source-dev-beam.sh:146` - The node-name host validator accepts unspecified IP literals like `0.0.0.0` or `::`, and those values are then fed into `ERL_EPMD_ADDRESS` and `inet_dist_use_interface`. That lets a misapplied gRPC-style wildcard host expose source-dev BEAM distribution broadly; reject unspecified/wildcard addresses in the shell bootstrap and the BEAM target parser.

🔧 Fix: Harden source-dev BEAM EPMD preflight
2 issues (1 error, 1 warning) still open:
- 🚨 `bin/lib/source-dev-beam.sh:319` - The EPMD preflight accepts an existing daemon as soon as one listener equals the requested host. A default `127.0.0.1` run will still pass if EPMD is also listening on a concrete non-loopback address like `10.0.0.2`, so the loopback source-dev mode can reuse an exposed EPMD. Reject listeners that are neither the requested host nor loopback, and add a regression with both `127.0.0.1` and an extra private IP in the fake `lsof` output.
- ⚠️ `bin/lib/source-dev-beam.sh:85` - The source-dev parser accepts IPv6 BEAM hosts, but the IEx launch args never add `-proto_dist inet6_tcp`. OTP documents IPv6 distribution as requiring that protocol, so `orchard_node_agent@::1` can be accepted into config and guardrails while the node still starts with the default IPv4 distribution. Decide whether this slice should reject IPv6 hosts or add the IPv6 distribution flag with coverage.

🔧 Fix: Harden source-dev BEAM IPv4 validation
✅ Re-checked - no issues remain.

</details>

<details>
<summary>✅ **Test** - passed</summary>

✅ No issues found.
- `bash scripts/test-source-dev-beam-bootstrap.sh`
- <code>`mise exec -- mix test apps/orchard_cli/test/orchard_cli/dev_scripts_test.exs apps/orchard_controller/test/orchard/config/source_dev_beam_test.exs apps/orchard_controller/test/orchard/config/runtime_target_parser_test.exs apps/orchard_shared/test/orchard/runtime_endpoint/domain_test.exs apps/orchard_node_agent/test/orchard/node/runtime_endpoint_mapper_test.exs apps/orchard_controller/test/orchard/inference_test.exs apps/orchard_controller/test/orchard/scheduler/multi_node_test.exs apps/orchard_controller/test/orchard/dispatch/probe_compatibility_test.exs apps/orchard_controller/test/orchard/console/runtime_test.exs apps/orchard_controller/test/orchard/console/nodes_live_test.exs` (initial setup attempt blocked by untrusted mise config)</code>
- `env MISE_TRUSTED_CONFIG_PATHS=/Users/najib/.no-mistakes/worktrees/2b20bc518d75/01KW24H36B7F8K1T9SSFQNY102/mise.toml mise exec -- mix test apps/orchard_cli/test/orchard_cli/dev_scripts_test.exs apps/orchard_controller/test/orchard/config/source_dev_beam_test.exs apps/orchard_controller/test/orchard/config/runtime_target_parser_test.exs apps/orchard_shared/test/orchard/runtime_endpoint/domain_test.exs apps/orchard_node_agent/test/orchard/node/runtime_endpoint_mapper_test.exs apps/orchard_controller/test/orchard/inference_test.exs apps/orchard_controller/test/orchard/scheduler/multi_node_test.exs apps/orchard_controller/test/orchard/dispatch/probe_compatibility_test.exs apps/orchard_controller/test/orchard/console/runtime_test.exs apps/orchard_controller/test/orchard/console/nodes_live_test.exs`
- `/var/folders/xy/bmzx3f853zd1tdvd11hgc0gm0000gn/T/no-mistakes-evidence/01KW24H36B7F8K1T9SSFQNY102/source-dev-beam-cli-evidence.sh`
- `env MISE_TRUSTED_CONFIG_PATHS=/Users/najib/.no-mistakes/worktrees/2b20bc518d75/01KW24H36B7F8K1T9SSFQNY102/mise.toml mise exec -- mix test --trace apps/orchard_controller/test/orchard/dispatch/probe_compatibility_test.exs:486 apps/orchard_controller/test/orchard/dispatch/probe_compatibility_test.exs:508 apps/orchard_controller/test/orchard/dispatch/probe_compatibility_test.exs:533 apps/orchard_controller/test/orchard/scheduler/multi_node_test.exs:610 apps/orchard_controller/test/orchard/scheduler/multi_node_test.exs:637 apps/orchard_controller/test/orchard/console/runtime_test.exs:1029 apps/orchard_controller/test/orchard/console/nodes_live_test.exs:811 apps/orchard_node_agent/test/orchard/node/runtime_endpoint_mapper_test.exs:50`

</details>

<details>
<summary>✅ **Document** - passed</summary>

✅ No issues found.

</details>

<details>
<summary>✅ **Lint** - passed</summary>

✅ No issues found.

</details>

<details>
<summary>✅ **Push** - passed</summary>

✅ No issues found.

</details>

### CODE REVIEW

File: bin/lib/source-dev-beam.sh [hunk 1]
```diff
@@ -0,0 +1,529 @@
...
+  if ! orchard_source_dev_beam_cookie_is_owner_only "$cookie_file"; then
+    echo "error: ORCHARD_BEAM_COOKIE_FILE must be owner-only" >&2
+    return 64
+  fi
+}
+
+orchard_source_dev_beam_cookie_is_owner_only() {
+  local cookie_file="$1"
+  local mode
+  mode="$(stat -f '%Lp' "$cookie_file" 2>/dev/null || stat -c '%a' "$cookie_file")"
```
[the remark this answers is not in the document]
Comment (najibninaba) [reply bin/lib/source-dev-beam.sh]: Confirmed. I reproduced this against PR head `84ffb75cb685da56767fc6fbb2cafbd2ea3b9372`.

Findings:
- `bin/lib/source-dev-beam.sh:492` still uses BSD-first `stat -f '%Lp'`.
- With GNU coreutils `stat` first on `PATH` and a `%Lp` sentinel path present, `stat -f '%Lp' "$cookie_file"` exits 0 with filesystem output, so `stat -c '%a'` never runs.
- In the launcher-shaped conditional path, a `0644` explicit cookie is accepted. `orchard_source_dev_beam_validate_cookie_file` returns success.
- `scripts/test-source-dev-beam-bootstrap.sh` covers ambient `0644` rejection, but does not fake GNU `stat -f` filesystem output.

I agree this should block merge. The fix should prefer GNU `stat -c '%a'` when available, fall back to BSD `stat -f '%Lp'`, validate mode output as octal digits only, and fail closed on unexpected output. We should also add a fake-tool regression for the GNU `stat -f` case before resolving this thread.

[the remark this answers is not in the document]
Comment (najibninaba) [reply bin/lib/source-dev-beam.sh]: Fixed in `cbb17db`.

What changed:
- `orchard_source_dev_beam_cookie_is_owner_only` now prefers GNU `stat -c '%a'`, falls back to BSD `stat -f '%Lp'`, accepts only octal mode output, and fails closed otherwise.
- `scripts/test-source-dev-beam-bootstrap.sh` now covers the fake GNU `stat -f` filesystem-output case and a non-octal stat-output failure path.

Validated:
- `bash -n bin/lib/source-dev-beam.sh scripts/test-source-dev-beam-bootstrap.sh bin/dev bin/dev-controller bin/dev-node-agent && scripts/test-source-dev-beam-bootstrap.sh`
- `mise exec -- mix test apps/orchard_cli/test/orchard_cli/dev_scripts_test.exs apps/orchard_controller/test/orchard/config/source_dev_beam_test.exs`

[the remark this answers is not in the document]
Comment (najibninaba) [reply bin/lib/source-dev-beam.sh]: @codex review

File: scripts/test-source-dev-beam-bootstrap.sh [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+    cat "$right" >&2
+    exit 1
+  fi
+}
+
+assert_mode() {
+  local expected="$1"
+  local path="$2"
+  local actual
+  actual="$(stat -f '%Lp' "$path" 2>/dev/null || stat -c '%a' "$path")"
```
[the remark this answers is not in the document]
Comment (najibninaba) [reply scripts/test-source-dev-beam-bootstrap.sh]: Fixed in `780aca9`.

What changed:
- `assert_mode` now prefers GNU `stat -c '%a'`, falls back to BSD `stat -f '%Lp'`, and fails closed on non-octal output.
- Added a focused fake-`stat` regression where `stat -c` returns `600` and `stat -f` returns filesystem-style text, so the old BSD-first helper would fail.

Validated:
- `bash -n bin/lib/source-dev-beam.sh scripts/test-source-dev-beam-bootstrap.sh bin/dev bin/dev-controller bin/dev-node-agent && scripts/test-source-dev-beam-bootstrap.sh`
- `mise exec -- mix test apps/orchard_cli/test/orchard_cli/dev_scripts_test.exs apps/orchard_controller/test/orchard/config/source_dev_beam_test.exs`

[the remark this answers is not in the document]
Comment (najibninaba) [reply scripts/test-source-dev-beam-bootstrap.sh]: @codex review
